flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/PostTransformOperator.java

```diff
@@ -362,25 +376,51 @@ private Optional<DataChangeEvent> processFilter(
     }
 
     private Optional<DataChangeEvent> processProjection(
-            TransformProjectionProcessor transformProjectionProcessor,
+            PostTransformProcessor postTransformProcessor,
             DataChangeEvent dataChangeEvent,
-            long epochTime)
-            throws Exception {
+            long epochTime) {
         BinaryRecordData before = (BinaryRecordData) dataChangeEvent.before();
         BinaryRecordData after = (BinaryRecordData) dataChangeEvent.after();
         if (before != null) {
             BinaryRecordData projectedBefore =
-                    transformProjectionProcessor.processData(before, epochTime);
+                    postTransformProcessor.processData(before, epochTime);
+            dataChangeEvent = DataChangeEvent.projectBefore(dataChangeEvent, projectedBefore);
+        }
+        if (after != null) {
+            BinaryRecordData projectedAfter = postTransformProcessor.processData(after, epochTime);
+            dataChangeEvent = DataChangeEvent.projectAfter(dataChangeEvent, projectedAfter);
+        }
+        return Optional.of(dataChangeEvent);
+    }
+
+    private Optional<DataChangeEvent> processPostProjection(
+            TableInfo tableInfo, DataChangeEvent dataChangeEvent) throws Exception {
+        BinaryRecordData before = (BinaryRecordData) dataChangeEvent.before();
+        BinaryRecordData after = (BinaryRecordData) dataChangeEvent.after();
+        if (before != null) {
+            BinaryRecordData projectedBefore = projectRecord(tableInfo, before);
             dataChangeEvent = DataChangeEvent.projectBefore(dataChangeEvent, projectedBefore);
         }
         if (after != null) {
-            BinaryRecordData projectedAfter =
-                    transformProjectionProcessor.processData(after, epochTime);
+            BinaryRecordData projectedAfter = projectRecord(tableInfo, after);
             dataChangeEvent = DataChangeEvent.projectAfter(dataChangeEvent, projectedAfter);
         }
         return Optional.of(dataChangeEvent);
     }
 
+    private BinaryRecordData projectRecord(TableInfo tableInfo, BinaryRecordData recordData) {
+        List<Object> valueList = new ArrayList<>();
+        RecordData.FieldGetter[] fieldGetters = tableInfo.getFieldGetters();
+
+        for (RecordData.FieldGetter fieldGetter : fieldGetters) {
+            valueList.add(fieldGetter.getFieldOrNull(recordData));
+        }
+
+        return tableInfo
+                .getRecordDataGenerator()
+                .generate(valueList.toArray(new Object[valueList.size()]));
+    }
+
     private boolean containFilteredComputedColumn(String projection, String filter) {
         boolean contain = false;
```

Comment:
Due to the filter always executing before the projection, this method is no longer useful.